File: frontend/mobile/lib/service/profile_service.dart
// lib/services/profile_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import '../models/profile_models.dart';
import '../config/api_config.dart';
import 'storage_helper.dart';
import 'auth_service.dart';

class ProfileService {
  // ========== ดึงข้อมูลโปรไฟล์ผู้ใช้ ==========
  static Future<UserProfile> getUserProfile(String userId) async {
    try {
      final url = '${ApiConfig.profileUrl}/$userId';

      final response = await AuthService.authenticatedRequest(
        method: 'GET',
        endpoint: url,
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> data = jsonDecode(response.body);
        return UserProfile.fromJson(data);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to fetch profile');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }

  // ========== ดึงข้อมูลโปรไฟล์ของ User ที่ login อยู่ ==========
  static Future<UserProfile> getMyProfile() async {
    try {
      final userId = await StorageHelper.getUserId();

      if (userId == null || userId.isEmpty) {
        throw Exception('No user ID found. Please login again.');
      }

      return await getUserProfile(userId);
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }

  // ========== อัปเดทรูปโปรไฟล์ ==========
  static Future<UpdateProfileImageResponse> updateProfileImage({
    required String userId,
    required File imageFile,
  }) async {
    try {
      String? accessToken = await StorageHelper.getAccessToken();

      if (accessToken == null || accessToken.isEmpty) {
        throw Exception('No access token found. Please login.');
      }

      final url = Uri.parse('${ApiConfig.profileUrl}/$userId/image');
      print('[Profile Service] Upload URL: $url');
      print('[Profile Service] Image path: ${imageFile.path}');
      print('[Profile Service] File exists: ${await imageFile.exists()}');

      // สร้าง multipart request
      var request = http.MultipartRequest('PUT', url);
      request.headers['Authorization'] = 'Bearer $accessToken';
      request.files.add(
        await http.MultipartFile.fromPath(
          'profile_image',
          imageFile.path,
        ),
      );

      print('[Profile Service] Sending request...');

      // ส่ง request
      var streamedResponse = await request.send();
      var response = await http.Response.fromStream(streamedResponse);

      print('[Profile Service] Response status: ${response.statusCode}');
      print('[Profile Service] Response body: ${response.body}');

      // ถ้า token หมดอายุ ให้ refresh แล้วลองใหม่
      if (response.statusCode == 401 || response.statusCode == 403) {
        print('[Profile Service] Token expired, refreshing...');
        try {
          accessToken = await AuthService.refreshAccessToken();
          print('[Profile Service] Token refreshed, retrying upload...');

          // ลองอัพโหลดอีกครั้งด้วย token ใหม่
          request = http.MultipartRequest('PUT', url);
          request.headers['Authorization'] = 'Bearer $accessToken';
          request.files.add(
            await http.MultipartFile.fromPath(
              'profile_image',
              imageFile.path,
            ),
          );

          streamedResponse = await request.send();
          response = await http.Response.fromStream(streamedResponse);

          print('[Profile Service] Retry response status: ${response.statusCode}');
          print('[Profile Service] Retry response body: ${response.body}');
        } catch (e) {
          print('[Profile Service] Refresh token error: $e');
          throw Exception('Session expired. Please login again.');
        }
      }

      if (response.statusCode == 200) {
        final Map<String, dynamic> data = jsonDecode(response.body);
        print('[Profile Service] Upload successful!');
        return UpdateProfileImageResponse(
          message: data['message'] ?? 'Profile image updated successfully',
          imageUrl: data['image_url'],
        );
      } else {
        // พยายาม parse error message จาก response
        String errorMessage = 'Failed to update profile image';
        try {
          final error = jsonDecode(response.body);
          errorMessage = error['message'] ?? errorMessage;
        } catch (e) {
          errorMessage = 'Server error: ${response.statusCode} - ${response.body}';
        }
        print('[Profile Service] Upload failed: $errorMessage');
        throw Exception(errorMessage);
      }
    } on SocketException catch (e) {
      print('[Profile Service] Network error: $e');
      throw Exception('ไม่สามารถเชื่อมต่อกับเซิร์ฟเวอร์ได้ กรุณาตรวจสอบการเชื่อมต่ออินเทอร์เน็ต');
    } catch (e) {
      print('[Profile Service] Error: $e');
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }

  // ========== อัปเดทข้อมูลโปรไฟล์ ==========
  static Future<UpdateProfileResponse> updateProfile({
    required String userId,
    double? weight,
    double? height,
    int? age,
    String? gender,
    String? goal,
  }) async {
    try {
      final url = '${ApiConfig.baseUrl}/api/update/$userId';

      final Map<String, dynamic> body = {};
      if (weight != null) body['weight'] = weight;
      if (height != null) body['height'] = height;
      if (age != null) body['age'] = age;
      if (gender != null) body['gender'] = gender.toLowerCase();
      if (goal != null) body['goal'] = goal.toLowerCase();

      final response = await AuthService.authenticatedRequest(
        method: 'PUT',
        endpoint: url,
        body: jsonEncode(body),
      );

      if (response.statusCode == 200) {
        final Map<String, dynamic> data = jsonDecode(response.body);
        return UpdateProfileResponse.fromJson(data);
      } else {
        final error = jsonDecode(response.body);
        throw Exception(error['message'] ?? 'Failed to update profile');
      }
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }

  // ========== อัปเดทโปรไฟล์ของ User ที่ login อยู่ ==========
  static Future<UpdateProfileResponse> updateMyProfile({
    double? weight,
    double? height,
    int? age,
    String? gender,
    String? goal,
  }) async {
    try {
      final userId = await StorageHelper.getUserId();

      if (userId == null || userId.isEmpty) {
        throw Exception('No user ID found. Please login again.');
      }

      return await updateProfile(
        userId: userId,
        weight: weight,
        height: height,
        age: age,
        gender: gender,
        goal: goal,
      );
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }

  // ========== อัปเดทรูปโปรไฟล์ของ User ที่ login อยู่ ==========
  static Future<UpdateProfileImageResponse> updateMyProfileImage({
    required File imageFile,
  }) async {
    try {
      // ✅ ดึง userId จาก JWT token โดยตรงแทนที่จะใช้จาก storage
      final accessToken = await StorageHelper.getAccessToken();

      if (accessToken == null || accessToken.isEmpty) {
        throw Exception('No access token found. Please login.');
      }

      String? userId;

      try {
        // Decode JWT payload เพื่อดึง userId
        final parts = accessToken.split('.');
        if (parts.length == 3) {
          final payload = parts[1];
          final normalized = base64Url.normalize(payload);
          final resp = utf8.decode(base64Url.decode(normalized));
          final payloadMap = json.decode(resp);

          // Backend ใช้ field 'id' ใน JWT
          userId = payloadMap['id']?.toString();

          print('[Profile Service] Decoded userId from token: $userId');
        }
      } catch (e) {
        print('[Profile Service] Error decoding token: $e');
      }

      // ถ้า decode ไม่ได้ ให้ fallback ไปใช้จาก storage
      if (userId == null || userId.isEmpty) {
        final userIdStr = await StorageHelper.getUserId();

        if (userIdStr == null || userIdStr.isEmpty) {
          throw Exception('User ID not found. Please login again.');
        }

        userId = userIdStr;
        print('[Profile Service] Using userId from storage: $userId');
      }

      return await updateProfileImage(
        userId: userId,
        imageFile: imageFile,
      );
    } catch (e) {
      if (e is Exception) rethrow;
      throw Exception('เกิดข้อผิดพลาด: ${e.toString()}');
    }
  }
}

File: frontend/web/src/app/services/profile_service.ts
// src/app/services/profile_service.ts

import axios from 'axios';

// ========================================
// Configuration
// ========================================
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:4000';

// ========================================
// Types
// ========================================
export interface UserProfile {
  user_id: number;
  username: string;
  email: string;
  image_profile?: string;
  image_profile_url?: string;
  phone_number?: string;
  age?: number;
  gender?: 'male' | 'female';
  height?: number;
  weight?: number;
  goal?: 'lose weight' | 'maintain weight' | 'gain weight';
  created_at?: string;
  updated_at?: string;
}

export interface UpdateProfileImageResponse {
  message: string;
  image_url: string;
}

export interface UpdateProfileData {
  age?: number;
  gender?: 'male' | 'female';
  height?: number;
  weight?: number;
  goal?: 'lose weight' | 'maintain weight' | 'gain weight';
}

export interface UpdateProfileResponse {
  message: string;
  user: UserProfile;
}

// ========================================
// Axios Instance
// ========================================
const profileAPI = axios.create({
  baseURL: API_BASE_URL,
  headers: {
    'Content-Type': 'application/json',
  },
  timeout: 10000,
});

// ========================================
// Request Interceptor (เพิ่ม token ทุก request)
// ========================================
profileAPI.interceptors.request.use(
  (config) => {
    const token = localStorage.getItem('accessToken'); // ✅ เปลี่ยนจาก 'token'
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => {
    return Promise.reject(error);
  }
);

// ========================================
// Response Interceptor (จัดการ error)
// ========================================
profileAPI.interceptors.response.use(
  (response) => response,
  (error) => {
    console.error('[Profile API] Error:', {
      status: error.response?.status,
      message: error.response?.data?.message,
      url: error.config?.url,
    });

    // Token หมดอายุ (401 Unauthorized)
    if (error.response?.status === 401 || error.response?.status === 403) {
      console.log('[Profile API] Token expired or forbidden, redirecting to login...');
      localStorage.removeItem('accessToken'); // ✅ เปลี่ยนจาก 'token'
      localStorage.removeItem('user');

      if (typeof window !== 'undefined') {
        window.location.href = '/login';
      }
    }

    // Network Error
    if (!error.response) {
      console.error('[Profile API] Network Error:', error.message);
    }

    return Promise.reject(error);
  }
);

// ========================================
// Profile API Services
// ========================================
export const profileService = {
  /**
   * ดึงข้อมูลโปรไฟล์ผู้ใช้ตาม ID
   */
  getUserProfile: async (userId: number): Promise<UserProfile> => {
    try {
      const response = await profileAPI.get<UserProfile>(`/api/profile/${userId}`);
      return response.data;
    } catch (error: any) {
      const errorMessage = error.response?.data?.message || 'เกิดข้อผิดพลาดในการดึงข้อมูลโปรไฟล์';
      throw new Error(errorMessage);
    }
  },

  /**
   * ดึงข้อมูลโปรไฟล์ของผู้ใช้ปัจจุบัน
   */
  getCurrentUserProfile: async (): Promise<UserProfile | null> => {
    try {
      const userStr = localStorage.getItem('user');
      if (!userStr) {
        console.warn('👤 [getCurrentUserProfile] User not found in localStorage');
        return null;
      }

      const user = JSON.parse(userStr);
      console.log('👤 [getCurrentUserProfile] User from localStorage:', user);

      // ตรวจสอบว่ามี id หรือไม่
      if (!user.id && !user.user_id) {
        console.error('❌ [getCurrentUserProfile] User ID not found in localStorage');
        return null;
      }

      const userId = user.id || user.user_id;
      console.log('👤 [getCurrentUserProfile] Fetching profile for user ID:', userId);

      const profile = await profileService.getUserProfile(userId);

      console.log('✅ [getCurrentUserProfile] Profile fetched:', profile);

      // อัปเดท localStorage ด้วยข้อมูลล่าสุด
      const updatedUser = {
        ...user,
        id: profile.user_id,
        username: profile.username,
        email: profile.email,
      };
      localStorage.setItem('user', JSON.stringify(updatedUser));

      console.log('✅ [getCurrentUserProfile] Updated user in localStorage:', updatedUser);

      return profile;
    } catch (error) {
      console.error('❌ [getCurrentUserProfile] Error:', error);
      return null;
    }
  },

  /**
   * อัปเดทข้อมูลโปรไฟล์
   */
  updateProfile: async (userId: number, data: UpdateProfileData): Promise<UpdateProfileResponse> => {
    try {
      const response = await profileAPI.put<UpdateProfileResponse>(
        `/api/update/${userId}`,
        data
      );

      // ✅ อัปเดท user ใน localStorage ด้วย
      const userStr = localStorage.getItem('user');
      if (userStr) {
        const currentUser = JSON.parse(userStr);
        const updatedUser = {
          ...currentUser,
          id: response.data.user.user_id,
          username: response.data.user.username,
          email: response.data.user.email,
          age: response.data.user.age,
          gender: response.data.user.gender,
          height: response.data.user.height,
          weight: response.data.user.weight,
          goal: response.data.user.goal,
        };
        localStorage.setItem('user', JSON.stringify(updatedUser));
      }

      return response.data;
    } catch (error: any) {
      const errorMessage = error.response?.data?.message || 'เกิดข้อผิดพลาดในการอัปเดทโปรไฟล์';
      throw new Error(errorMessage);
    }
  },

  /**
   * อัปเดทรูปโปรไฟล์
   */
  updateProfileImage: async (userId: number, imageFile: File): Promise<UpdateProfileImageResponse> => {
    try {
      const formData = new FormData();
      formData.append('profile_image', imageFile);

      const response = await profileAPI.put<UpdateProfileImageResponse>(
        `/api/profile/${userId}/image`,
        formData,
        {
          headers: {
            'Content-Type': 'multipart/form-data',
          },
        }
      );

      return response.data;
    } catch (error: any) {
      const errorMessage = error.response?.data?.message || 'เกิดข้อผิดพลาดในการอัปเดทรูปโปรไฟล์';
      throw new Error(errorMessage);
    }
  },

  /**
   * ดึง URL รูปโปรไฟล์เต็ม
   */
  getProfileImageUrl: (imageName?: string): string => {
    if (!imageName) {
      return '/pic/person.png'; // รูป default
    }
    return `${API_BASE_URL}/uploads/${imageName}`;
  },
};

export default profileService;